File: python/bKap3_1_InSampleRegression.py
"""
bKap3_1_InSampleRegression.py
Python port of bKap3_1_InSampleRegression.m.

In-sample predictive regression of inflation on a single (z-standardised)
predictor, with intercept. Reports beta, t-statistics and R2.

Run from the repo root:  python python/bKap3_1_InSampleRegression.py
"""

import numpy as np
import pandas as pd

import models

# =========================================================================
#  Load data
# =========================================================================
DATA_PATH = "./DATA/Liedtke/US/aggregated.csv"
tData = pd.read_csv(DATA_PATH)
cAllNames = list(tData.columns)
dtDates = pd.to_datetime(tData.iloc[:, 0])

mX = tData.iloc[:, 2:].to_numpy(float)     # predictors
vY = tData.iloc[:, 1].to_numpy(float)      # inflation target
cXnamesM = cAllNames[2:]

# Predictor of interest (default: first predictor)
sPred = cXnamesM[0]
iIdxPred = cXnamesM.index(sPred)
vX = mX[:, iIdxPred]

# =========================================================================
#  Data preprocessing
# =========================================================================
iTimeLag = 1
# Lag the predictor so it is known one period ahead of the target
vXlag = np.full_like(vX, np.nan)
vXlag[iTimeLag:] = vX[:-iTimeLag]

# Remove missing values
lIsNaN = np.isnan(vXlag) | np.isnan(vY)
keep = ~lIsNaN
vYk = vY[keep]
vXk = vXlag[keep]
dtDates = dtDates[keep].reset_index(drop=True)

# z-transformation of predictor (sample std, ddof=1)
vXz = (vXk - vXk.mean()) / vXk.std(ddof=1)

# =========================================================================
#  In-sample regression with intercept
# =========================================================================
reg = models.ols_regress(vYk, vXz)
vBeta = reg["beta"]      # [intercept, slope]
vBetaT = reg["t"]
dR2 = reg["R2"]

print(f"In-sample regression: inflation ~ {sPred} (z-standardised)")
print(f"  Intercept : {vBeta[0]:10.6f} (t = {vBetaT[0]:7.3f})")
print(f"  Beta      : {vBeta[1]:10.6f} (t = {vBetaT[1]:7.3f})")
print(f"  R2        : {dR2:10.6f}")

File: bKap3_1_InSampleRegression.m
% Script for performing the in-sample regressions in GWZ

% Clear console
clear; clc; close all;

% Set paths
sOldPath = path;
sDataPath = './DATA/Liedtke/US/';
sResultsPath = './RESULTS/GWZ/';
addpath('./Utils/');

%% Load data
% Macro panel produced by DATA/Liedtke/aggregate.py.
% CSV layout: col 1 = observation_date, col 2 = target (inflation),
%             col 3..end = predictors (already lag-aligned in Python).
tData     = readtable([sDataPath, 'aggregated.csv']);
cAllNames = tData.Properties.VariableNames;

% Parse dates (first column)
dtDates = tData{:,1};
if ~isdatetime(dtDates)
    dtDates = datetime(dtDates);
end

% Target (col 2) and predictors (col 3..end)
mX          = tData{:,3:end};          % predictors
vY          = tData{:,2};              % inflation target
cXnamesM    = cAllNames(3:end)';       % predictor names as an Nx1 cell

% Find predictor of interest (edit sPred to any predictor column name)
sPred       = cXnamesM{1};             % default: first predictor
iIdxPred    = find(strcmpi(cXnamesM,sPred));

% Get the predictor
vX          = mX(:,iIdxPred);

%% Data preprocessing
% Predictive lag (periods). The predictors are already reporting-lag aligned
% in Python; this is the additional one-period predictive lag.
iTimeLag = 1;

% Determine dimensions
[iNumObs, iNumAssets]       = size(vY);
[iNumObsP, iNumPredictors]  = size(vX);

% Lag the predictor so that it is known one period ahead of the target
vXlag = [NaN(iTimeLag,iNumPredictors); vX(1:end-iTimeLag,:)];

%% In-Sample analysis
% Remove missing values
lIsNaN          = any(isnan(vXlag),2) | isnan(vY);
vY(lIsNaN,:)    = [];
vXlag(lIsNaN,:) = [];
dtDates(lIsNaN) = [];

% z-Transformation of predictor
vXlag = (vXlag - mean(vXlag))./std(vXlag);

% Regression with intercept
rResults = regstats2(vY, vXlag, 'linear', {'tstat','rsquare'});
vBeta    = rResults.tstat.beta;      % coefficient estimates
vBetaT   = rResults.tstat.t;         % t-Statistic estimates
dR2      = rResults.rsquare;         % R2

% Restore path
path(sOldPath);